apis/droplet_handler_test.go

```diff
@@ -0,0 +1,195 @@
+						GUID:      dropletGUID,
+						State:     "STAGED",
+						CreatedAt: createdAt,
+						UpdatedAt: updatedAt,
+						Lifecycle: repositories.Lifecycle{
+							Type: "buildpack",
+							Data: repositories.LifecycleData{
+								Buildpacks: []string{},
+								Stack:      "",
+							},
+						},
+						Stack: "cflinuxfs3",
+						ProcessTypes: map[string]string{
+							"rake": "bundle exec rake",
+							"web":  "bundle exec rackup config.ru -p $PORT",
+						},
+						AppGUID:     appGUID,
+						PackageGUID: packageGUID,
+					}, nil)
+					router.ServeHTTP(rr, req)
+				})
+				It("returns status 200 OK", func() {
+					Expect(rr.Code).To(Equal(http.StatusOK), "Matching HTTP response code:")
+				})
+				It("returns Content-Type as JSON in header", func() {
+					contentTypeHeader := rr.Header().Get("Content-Type")
+					Expect(contentTypeHeader).To(Equal(jsonHeader), "Matching Content-Type header:")
+				})
+				It("configures the client", func() {
+					Expect(clientBuilder.CallCount()).To(Equal(1))
+				})
+				It("fetches the right droplet", func() {
+					Expect(dropletRepo.FetchDropletCallCount()).To(Equal(1))
+
+					_, _, actualDropletGUID := dropletRepo.FetchDropletArgsForCall(0)
+					Expect(actualDropletGUID).To(Equal(dropletGUID))
+				})
+				It("returns the droplet in the response", func() {
+					Expect(rr.Body.String()).To(MatchJSON(`{
+					  "guid": "`+dropletGUID+`",
+					  "state": "STAGED",
+					  "error": null,
+					  "lifecycle": {
+						"type": "buildpack",
+						"data": {
+							"buildpacks": [],
+							"stack": ""
+						}
+					},
+					  "execution_metadata": "",
+					  "process_types": {
+						"rake": "bundle exec rake",
+						"web": "bundle exec rackup config.ru -p $PORT"
+					  },
+					  "checksum": null,
+					  "buildpacks": [],
+					  "stack": "cflinuxfs3",
+					  "image": null,
+					  "created_at": "`+createdAt+`",
+					  "updated_at": "`+updatedAt+`",
+					  "relationships": {
+						"app": {
+						  "data": {
+							"guid": "`+appGUID+`"
+						  }
+						}
+					  },
+					  "links": {
+						"self": {
+						  "href": "`+defaultServerURI("/v3/droplets/", dropletGUID)+`"
+						},
+						"package": {
+						  "href": "`+defaultServerURI("/v3/packages/", packageGUID)+`"
+						},
+						"app": {
+						  "href": "`+defaultServerURI("/v3/apps/", appGUID)+`"
+						},
+						"assign_current_droplet": {
+						  "href": "`+defaultServerURI("/v3/apps/", appGUID, "/relationships/current_droplet")+`",
+						  "method": "PATCH"
+						  },
+						"download": null
+					  },
+					  "metadata": {
+						"labels": {},
+						"annotations": {}
+					  }
+					}`), "Response body matches response:")
+				})
+			})
+		})
+		When("building the k8s client errors", func() {
+			BeforeEach(func() {
+				clientBuilder.Returns(nil, errors.New("boom"))
+				router.ServeHTTP(rr, req)
+			})
+
+			itRespondsWithUnknownError(getRR)
+		})
+		When("build staging fails", func() {
```

Comment:
Since we're abstracting how the Droplet is represented with the DropletRepo, I'd rephrase this to "the droplet doesn't exist". That's more direct.

Reply:
Actually, now that I've read further I think you should just delete this case and the one below about "staging is not complete". The setup for all 3 is identical and it doesn't matter *WHY* you're in that scenario, just that you are.